Log any FQDN lookup errors and fallback to OS-reported hostname

ycombinator opened the pull request:
<!-- Type of change
Please label this PR with one of the following labels, depending on the scope of your change:
- Bug
- Enhancement
- Breaking change
- Deprecation
- Cleanup
- Docs
-->

## What does this PR do?

This PR fixes a bug wherein a Beat would fail to start if the FQDN lookup failed.

<!-- Mandatory
Explain here the changes you made on the PR. Please explain the WHAT: patterns used, algorithms implemented, design architecture, message processing, etc.
-->

## Why is it important?

FQDN lookup is "best effort".  As such, if the lookup fails for some reason, we should not fail to start a Beat or halt execution otherwise.  Instead, we should fallback to the OS-provided hostname and continue execution.

<!-- Mandatory
Explain here the WHY, or the rationale/motivation for the changes.
-->

## Checklist

<!-- Mandatory
Add a checklist of things that are required to be reviewed in order to have the PR approved

List here all the items you have verified BEFORE sending this PR. Please DO NOT remove any item, striking through those that do not apply. (Just in case, strikethrough uses two tildes. ~~Scratch this.~~)
-->

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] ~I have made corresponding changes to the documentation~
- [ ] ~I have made corresponding change to the default configuration files~
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] ~I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.~

## How to test this PR locally

<!-- Recommended
Explain here how this PR will be tested by the reviewer: commands, dependencies, steps, etc.
-->

1. Build Filebeat (or any other Beat) with this PR.
   ```sh
   cd beats/filebeat
   mage clean build
   ```

2. Set hostname to something that wouldn't resolve via DNS.  On MacOS this can be done via:
   ```sh
   orig_hostname=$(hostname -f)
   sudo scutil --set HostName cowchicken
   ```

3. Run `filebeat version`.  This command should report the version as usual and NOT fail with an FQDN-related error (as reported in #34910)
   ```sh
   ./filebeat version
   ```
   ```
   filebeat version 8.8.0 (arm64), libbeat 8.8.0 [528128f44ee870cedca830d8acbf63d8ee35ae26 built 2023-03-28 16:45:09 +0000 UTC]
   ```

4. Create a configuration file for the Beat that enables the `add_host_metadata` processor and also enables the FQDN feature flag.
   ```sh
   cat <<EOF >filebeat.test.yml
   filebeat.inputs:
   - type: stdin
   
   processors:
   - add_host_metadata: ~
   
   features.fqdn.enabled: true
   output.console.enabled: true
   EOF
   ```

5.  Start the Beat with the above configuration and look for log entries mentioning `FQDN`.  You should see two log entries in all, one from Beat initialization and one from the `add_host_metadata` processor.

    ```sh
    ./filebeat -c ./filebeat.test.yml -e 2>&1 | grep -i fqdn
    ```

    ```
    {"log.level":"warn","@timestamp":"2023-03-28T11:38:17.639-0700","log.origin":{"file.name":"instance/beat.go","file.line":802},"message":"unable to lookup FQDN: could not get FQDN, all methods failed: failed looking up CNAME: lookup cowchicken: no such host: failed looking up IP: lookup cowchicken: no such host, using hostname = cowchicken as FQDN","service.name":"filebeat","ecs.version":"1.6.0"}
    {"log.level":"warn","@timestamp":"2023-03-28T11:38:17.642-0700","log.logger":"add_host_metadata","log.origin":{"file.name":"add_host_metadata/add_host_metadata.go","file.line":152},"message":"unable to lookup FQDN: could not get FQDN, all methods failed: failed looking up CNAME: lookup cowchicken: no such host: failed looking up IP: lookup cowchicken: no such host, using hostname = cowchicken as FQDN","service.name":"filebeat","ecs.version":"1.6.0"}
    ```

99. Reset hostname at the end of the test.  On MacOS this can be done via:
    ```sh
    sudo scutil --set HostName $orig_hostname
    ```

## Related issues

<!-- Recommended
Link related issues below. Insert the issue link or reference after the word "Closes" if merging this should automatically close it.

- Closes #123
- Relates #123
- Requires #123
- Superseds #123
-->
- Closes #34910
- Follow up to https://github.com/elastic/beats/pull/34456

## Use cases

<!-- Recommended
Explain here the different behaviors that this PR introduces or modifies in this project, user roles, environment configuration, etc.

If you are familiar with Gherkin test scenarios, we recommend its usage: https://cucumber.io/docs/gherkin/reference/
-->

## Screenshots

<!-- Optional
Add here screenshots about how the project will be changed after the PR is applied. They could be related to web pages, terminal, etc, or any other image you consider important to be shared with the team.
-->

## Logs

<!-- Recommended
Paste here output logs discovered while creating this PR, such as stack traces or integration logs, or any other output you consider important to be shared with the team.
-->

elasticmachine commented:
## :broken_heart: Build Failed


<!-- BUILD BADGES-->
> _the below badges are clickable and redirect to their specific view in the CI or DOCS_
[![Pipeline View](https://img.shields.io/badge/pipeline-pipeline%20-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-34946/detail/PR-34946/1//pipeline) [![Test View](https://img.shields.io/badge/test-test-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-34946/detail/PR-34946/1//tests) [![Changes](https://img.shields.io/badge/changes-changes-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-34946/detail/PR-34946/1//changes) [![Artifacts](https://img.shields.io/badge/artifacts-artifacts-yellow)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-34946/detail/PR-34946/1//artifacts) [![preview](https://img.shields.io/badge/docs-preview-yellowgreen)](http://beats_34946.docs-preview.app.elstc.co/diff)  [![preview](https://img.shields.io/badge/elastic-observability-blue)](https://ci-stats.elastic.co/app/apm/services/beats-ci/transactions/view?rangeFrom=2023-03-28T00:14:59.993Z&rangeTo=2023-03-28T00:34:59.993Z&transactionName=BUILD+Beats%2Fbeats%2FPR-%7Bnumber%7D&transactionType=job&latencyAggregationType=avg&traceId=ad20348bc600ebcf3a433bd0cd12bf84&transactionId=4c9977039a623325)

<!-- BUILD SUMMARY-->
<details><summary>Expand to view the summary</summary>
<p>

#### Build stats




* Start Time: 2023-03-28T00:24:59.993+0000


* Duration: 63 min 38 sec




#### Test stats :test_tube:

| Test         | Results                         |
| ------------ | :-----------------------------: |
| Failed       | 0  |
| Passed       | 5759  |
| Skipped      | 161 |
| Total        | 5920   |


</p>
</details>

<!-- TEST RESULTS IF ANY-->


<!-- STEPS ERRORS IF ANY -->














### Steps errors [![181](https://img.shields.io/badge/181%20-red)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-34946/detail/PR-34946/1//pipeline)
  <details><summary>Expand to view the steps failures</summary>
  <p>


> Show only the first 10 steps failures

  
  
  
  
##### `x-pack/winlogbeat-build-windows-2019 - mage build unitTest`
<ul>
<li>Took 2 min 56 sec . View more details <a href="https://beats-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/Beats/pipelines/beats/pipelines/PR-34946/runs/1/steps/13681/log/?start=0">here</a></li>
<li>Description: <code>mage build unitTest</code></l1>
</ul>
  
  
  
  
##### `x-pack/winlogbeat-build-windows-2019 - mage build unitTest`
<ul>
<li>Took 0 min 28 sec . View more details <a href="https://beats-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/Beats/pipelines/beats/pipelines/PR-34946/runs/1/steps/15115/log/?start=0">here</a></li>
<li>Description: <code>mage build unitTest</code></l1>
</ul>
  
  
  
  
##### `x-pack/winlogbeat-build-windows-2019 - mage build unitTest`
<ul>
<li>Took 1 min 30 sec . View more details <a href="https://beats-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/Beats/pipelines/beats/pipelines/PR-34946/runs/1/steps/15486/log/?start=0">here</a></li>
<li>Description: <code>mage build unitTest</code></l1>
</ul>
  
  
  
  
##### `x-pack/winlogbeat-windows-2022-windows-2022 - mage build unitTest`
<ul>
<li>Took 2 min 2 sec . View more details <a href="https://beats-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/Beats/pipelines/beats/pipelines/PR-34946/runs/1/steps/12811/log/?start=0">here</a></li>
<li>Description: <code>mage build unitTest</code></l1>
</ul>
  
  
  
  
##### `x-pack/winlogbeat-windows-2022-windows-2022 - mage build unitTest`
<ul>
<li>Took 0 min 26 sec . View more details <a href="https://beats-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/Beats/pipelines/beats/pipelines/PR-34946/runs/1/steps/14582/log/?start=0">here</a></li>
<li>Description: <code>mage build unitTest</code></l1>
</ul>
  
  
  
  
##### `x-pack/winlogbeat-windows-2022-windows-2022 - mage build unitTest`
<ul>
<li>Took 1 min 30 sec . View more details <a href="https://beats-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/Beats/pipelines/beats/pipelines/PR-34946/runs/1/steps/14925/log/?start=0">here</a></li>
<li>Description: <code>mage build unitTest</code></l1>
</ul>
  
  
  
  
##### `x-pack/winlogbeat-windows-2016-windows-2016 - mage build unitTest`
<ul>
<li>Took 2 min 52 sec . View more details <a href="https://beats-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/Beats/pipelines/beats/pipelines/PR-34946/runs/1/steps/13430/log/?start=0">here</a></li>
<li>Description: <code>mage build unitTest</code></l1>
</ul>
  
  
  
  
##### `x-pack/winlogbeat-windows-2016-windows-2016 - mage build unitTest`
<ul>
<li>Took 0 min 25 sec . View more details <a href="https://beats-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/Beats/pipelines/beats/pipelines/PR-34946/runs/1/steps/14891/log/?start=0">here</a></li>
<li>Description: <code>mage build unitTest</code></l1>
</ul>
  
  
  
  
##### `x-pack/winlogbeat-windows-2016-windows-2016 - mage build unitTest`
<ul>
<li>Took 0 min 29 sec . View more details <a href="https://beats-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/Beats/pipelines/beats/pipelines/PR-34946/runs/1/steps/15228/log/?start=0">here</a></li>
<li>Description: <code>mage build unitTest</code></l1>
</ul>
  
  
  
  
##### `Error signal`
<ul>
<li>Took 0 min 0 sec . View more details <a href="https://beats-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/Beats/pipelines/beats/pipelines/PR-34946/runs/1/steps/15225/log/?start=0">here</a></li>
<li>Description: <code>Error 'hudson.AbortException: script returned exit code 2'</code></l1>
</ul>
  
  </p>
  </details>



## :green_heart: Flaky test report
Tests succeeded.



## :robot: GitHub comments

<details><summary>Expand to view the GitHub comments</summary>
<p>

To re-run your PR in the CI, just comment with:



- `/test` : Re-trigger the build.


- `/package` : Generate the packages and run the E2E tests.


- `/beats-tester` : Run the installation tests with beats-tester.


- `run` `elasticsearch-ci/docs` : Re-trigger the docs validation. (use unformatted text in the comment!)




</p>
</details>

<!--COMMENT_GENERATED_WITH_ID_comment.id-->

ycombinator commented on libbeat/processors/add_host_metadata/add_host_metadata.go:
```diff
@@ -143,7 +143,19 @@ func (p *addHostMetadata) loadData() error {
+		}
+	}
+
+	data := host.MapHostInfo(h.Info(), hostname)
```
Depends on https://github.com/elastic/elastic-agent-system-metrics/pull/81 being merged and dependency on `elastic-agent-system-metrics` being bumped up.

ycombinator commented on libbeat/cmd/instance/beat.go:
```diff
@@ -595,7 +602,7 @@ func (b *Beat) RegisterHostname(useFQDN bool) {
 	// state.host
 	stateRegistry := monitoring.GetNamespace("state").GetRegistry()
-	monitoring.NewFunc(stateRegistry, "host", host.ReportInfo(useFQDN), monitoring.Report)
+	monitoring.NewFunc(stateRegistry, "host", host.ReportInfo(hostname), monitoring.Report)
```
Depends on https://github.com/elastic/elastic-agent-system-metrics/pull/81 being merged and dependency on `elastic-agent-system-metrics` being bumped up.

ycombinator commented on libbeat/cmd/instance/beat.go:
```diff
@@ -248,7 +248,14 @@ func NewBeat(name, indexPrefix, v string, elasticLicensed bool) (*Beat, error) {
+	if err != nil {
+		// FQDN lookup is "best effort".  We log the error, fallback to
+		// the OS-reported hostname, and move on.
+		logp.NewLogger(name).Infof("unable to lookup FQDN: %s, using hostname = %s", err.Error(), hostname)
```
I'm not able to get this log message to be printed to the logs.  After reading the godoc for the `logp.NewLogger` function, I've also tried using `logp.L()` instead but that didn't work either.  Adding a `fmt.Println` right above this line works, though.  @faec is helping me figure out why the logging might not be working here.

ycombinator replied:
Turns out logging isn't configured at this point so we need to move the FQDN initialization to slightly later in the execution.

ycombinator commented on libbeat/cmd/instance/beat.go:
```diff
@@ -244,12 +244,6 @@ func NewBeat(name, indexPrefix, v string, elasticLicensed bool) (*Beat, error) {
-		return nil, fmt.Errorf("failed to get host information: %w", err)
-	}
-	fqdn := h.Info().FQDN
-
```
This piece of code had to be moved to slightly later in the Beat's execution because we need logging to be configured to emit a warning if the FQDN lookup fails.  You'll find it elsewhere in this PR, along with the logging.

elasticmachine commented:
Pinging @elastic/elastic-agent (Team:Elastic-Agent)

cmacknz commented on libbeat/processors/add_host_metadata/add_host_metadata.go:
```diff
@@ -143,7 +143,19 @@ func (p *addHostMetadata) loadData() error {
+		if err != nil {
+			// FQDN lookup is "best effort".  We log the error, fallback to
+			// the OS-reported hostname, and move on.
+			p.logger.Warnf("unable to lookup FQDN: %s, using hostname = %s as FQDN", err.Error(), hostname)
```
How often would this be logged? We may want to lower this to Debug to avoid unnecessarily spamming the logs when this fails.

cmacknz replied:
It feels like we would be better off with a metrics counter for how often this resolution has failed.

That would tell us it's failing without worrying about log spam, and we could then turn on debug logging to see exactly what is happening.

ycombinator replied:
> It feels like we would be better off with a metrics counter for how often this resolution has failed.
> 
> That would tell us it's failing without worrying about log spam, and we could then turn on debug logging to see exactly what is happening.

Makes sense.  Let me look into implementing a metrics counter.

cmacknz commented:
I assume writing a regression test for this is challenging because it requires us to modify the hostname of the system running the tests to one that doesn't resolve via DNS?

cmacknz commented:
Also your test steps are excellent, thank you! I followed them exactly and the lack of fqdn logs caught that my attempt to switch branches failed because I had local file modifications to discard first :p

ycombinator commented:
> I assume writing a regression test for this is challenging because it requires us to modify the hostname of the system running the tests to one that doesn't resolve via DNS?

Yes. Perhaps we can do something with a Docker testcontainer, though.  Let me look into that.

leehinman commented:
> > I assume writing a regression test for this is challenging because it requires us to modify the hostname of the system running the tests to one that doesn't resolve via DNS?
> 
> Yes. Perhaps we can do something with a Docker testcontainer, though. Let me look into that.

maybe https://github.com/foxcpp/go-mockdns ?

ycombinator commented:
Thanks to @leehinman's suggestion about https://github.com/foxcpp/go-mockdns, I was able to implement a couple of unit test cases - one where the FQDN lookup will succeed and one where it will fail.  I've also implemented the metrics counter and dropped the log level for the FQDN lookup failure in the `add_host_metadata` processor to `debug`, as suggested by @cmacknz.

This PR is ready for review again.

cmacknz commented on libbeat/processors/add_host_metadata/add_host_metadata.go:
```diff
@@ -48,23 +63,30 @@ type addHostMetadata struct {
 
+	// Logging and metrics (each processor instance has a unique ID).
+	var (
+		id     = int(instanceID.Inc())
```
First time I've seen this way of getting around the namespace collision issue. CC @fearful-symmetry, this could help the version of this problem you have in the shipper.

fearful-symmetry replied:
huh, interesting!

cmacknz commented on libbeat/processors/add_host_metadata/add_host_metadata.go:
```diff
@@ -48,23 +63,30 @@ type addHostMetadata struct {
+	var (
+		id     = int(instanceID.Inc())
+		logger = logp.NewLogger(logName).With("instance_id", id)
+		reg    = monitoring.Default.NewRegistry(logName+"."+strconv.Itoa(id), monitoring.DoNotReport)
```
Do we actually need a separate instance of this per processor? Do we expect the errors to be unique per processor instance? There's also no way to map the ID here back to the configuration, so it seems a bit arbitrary and is probably simpler if we just have one atomic counter shared by all instances of this processor.

```json
      "processor": {
        "add_host_metadata": {
          "1": {
            "fqdn_lookup_failed": 1
          }
        }
      },
```

99% of the time there is just going to be the one add_host_metadata instance applied to each event collected since that is what is configured by default. If someone configured this to be per input we could end up with a lot of processor instances (think k8s autodiscovery where every pod gets its own input).

ycombinator replied:
Fixed via d9d55c7735.

ycombinator commented:
/test
